### DISCUSSION TITLE
VoLTE support and sound quality

### DISCUSSION THREAD

**Discussion by natan2525** [IchthysMaranatha/asterisk-chan-quectel#47] [Q&A]
Can anyone share what's the expected call quality with EC25-E or any other quectel module?
Technically docs say it supports VoLTE (_optional?_), but UAC documentation specifies only 8KHz 16bit audio over USB. 

VoLTE support is becoming a must as providers are slowly disabling 3G networks in the coming years, but if voice quality is supbar i'd consider buying an actual VoIP gateway instead.

I've already experienced bad quality with chan_dongle and old huawei modem, where people were actively telling me the call quality is worse (compared to smartphone on 3G) and I don't want to repeat that.

**Comment by spacedream** [comment]
Actually sound quality can be bad in many reasons. One of them is not tuned well network/kernel/etc on voip server. Or even its can be done quality degrade by third-party transit provider, as this (darky/greedy) technologies exists for many years already. Test quality on your local cluster/setup first as mandatory.   


May 21, 2023, 13:10 by ***@***.***:

**Comment by RoEdAl** [comment]
Currently SimCOM *SIM7600G-H* module supports 16kHz audio over USB (serial port).
Unfortunately, I don't know how to enable *VoLTE* and 16kHz mode does not improve audio quality at all.

One may try some `AT` commands to improve audio quality. See my [wiki page](//github.com/RoEdAl/asterisk-chan-quectel/wiki/Improving-audio-quality) for more info.

**Comment by basncy** [comment]
is it possible to port to android termux?(with minimal asterisk build)
in my case, AT command: echo 'AT+CSQ\r' >/dev/at_mdm0
sound card: cat /proc/asound/card0/id
msm8998tavilsnd
adb command logcat -s audio_route ActivityManager also help if we want to control im apps(by send input in adb shell)

**Comment by hellkaim** [comment]
I generally had better results with LTE vs GSM despite NOT USING UAC.
Another issue that not every modem will give you VoLTE and on top of it not all sim cards with that particular module will register VoLTE. I found that to be the case in Spain with Orange.

Despite that, over all voice quality is really bad but I think that may be tuned as I saw fellow with several LTE modems being happy about how it works.

**Comment by humuu135** [comment]
When I executed `AT+QAUDCFG=?` on my EC25-J, the response included `+QAUDCFG: "uac_fs",(0-1)`.
When I tried the following command, I was able to set UAC to 16kHz.

`AT+QAUDCFG="uac_fs",1`

(After rebooting the device)
```
# cat /proc/asound/card0/pcm0p/sub0/hw_params
access: MMAP_INTERLEAVED
format: S16_LE
subformat: STD
channels: 1
rate: 16000 (16000/1)
period_size: 320
buffer_size: 16000
```


Also, when I executed the following command on my EC25-J, the response was `+QNVFR: 01000000` and it seemed that 6.6kbps was set for AMR-WB.

`AT+QNVFR="/nv/item_files/ims/qipcall_codec_mode_set_amr_wb"`

When I tried the following command, the 5kHz test tone played by aplay can now be heard by the other party.

`AT+QNVFW="/nv/item_files/ims/qipcall_codec_mode_set_amr_wb",00000000`

(Reference: https://blog.csdn.net/ZhongGuoRenMei/article/details/108859123)

Both commands took effect after rebooting the device. (`AT+QPOWD=1`)

I could not find a channel driver that supports UAC16kHz for Quectel EC25, so I would appreciate it if someone could implement it.
